File: app/src/main/java/com/metehanyl/calarsaat/alarm/AlarmSounds.kt
package com.metehanyl.calarsaat.alarm

import com.metehanyl.calarsaat.R

/** A single tone segment. frequencyHz <= 0 means silence for durationMs. */
data class AlarmTone(val frequencyHz: Int, val durationMs: Int)

data class AlarmMelody(
    val id: Int,
    val nameRes: Int,
    val tones: List<AlarmTone>,
    val tailSilenceMs: Int
)

/** Five procedurally synthesized, alarm-appropriate tone melodies (no bundled audio files needed). */
object AlarmSounds {
    val ALL = listOf(
        AlarmMelody(
            id = 0,
            nameRes = R.string.melody_classic,
            tones = listOf(
                AlarmTone(1000, 150), AlarmTone(0, 80),
                AlarmTone(1000, 150), AlarmTone(0, 80),
                AlarmTone(750, 220)
            ),
            tailSilenceMs = 350
        ),
        AlarmMelody(
            id = 1,
            nameRes = R.string.melody_digital,
            tones = listOf(
                AlarmTone(1300, 90), AlarmTone(0, 90),
                AlarmTone(1300, 90), AlarmTone(0, 90),
                AlarmTone(1300, 90), AlarmTone(0, 90),
                AlarmTone(1300, 90)
            ),
            tailSilenceMs = 400
        ),
        AlarmMelody(
            id = 2,
            nameRes = R.string.melody_triplet,
            tones = listOf(
                AlarmTone(800, 130), AlarmTone(0, 60),
                AlarmTone(900, 130), AlarmTone(0, 60),
                AlarmTone(1050, 130)
            ),
            tailSilenceMs = 450
        ),
        AlarmMelody(
            id = 3,
            nameRes = R.string.melody_rising,
            tones = listOf(
                AlarmTone(600, 90), AlarmTone(750, 90),
                AlarmTone(900, 90), AlarmTone(1050, 90),
                AlarmTone(1200, 110)
            ),
            tailSilenceMs = 400
        ),
        AlarmMelody(
            id = 4,
            nameRes = R.string.melody_siren,
            tones = listOf(
                AlarmTone(1100, 300), AlarmTone(750, 300),
                AlarmTone(1100, 300), AlarmTone(750, 300)
            ),
            tailSilenceMs = 250
        )
    )

    fun byId(id: Int): AlarmMelody = ALL.firstOrNull { it.id == id } ?: ALL[0]
}

File: ios/CalarSaat/Audio/AlarmSounds.swift
import Foundation

struct AlarmTone {
    /// Hz; <= 0 means silence.
    let frequencyHz: Int
    let durationMs: Int
}

struct AlarmMelody: Identifiable {
    let id: Int
    let nameKey: String
    let tones: [AlarmTone]
    let tailSilenceMs: Int

    /// Filename (without extension) of the pre-rendered ~28s loop bundled as a
    /// notification sound. Must match `Resources/Sounds/<fileBaseName>.wav`.
    var fileBaseName: String { nameKey }
}

/// Mirrors Android's `AlarmSounds` melody table exactly (frequencies/durations).
enum AlarmSounds {
    static let all: [AlarmMelody] = [
        AlarmMelody(
            id: 0,
            nameKey: "melody_classic",
            tones: [
                AlarmTone(frequencyHz: 1000, durationMs: 150),
                AlarmTone(frequencyHz: 0, durationMs: 80),
                AlarmTone(frequencyHz: 1000, durationMs: 150),
                AlarmTone(frequencyHz: 0, durationMs: 80),
                AlarmTone(frequencyHz: 750, durationMs: 220),
            ],
            tailSilenceMs: 350
        ),
        AlarmMelody(
            id: 1,
            nameKey: "melody_digital",
            tones: [
                AlarmTone(frequencyHz: 1300, durationMs: 90),
                AlarmTone(frequencyHz: 0, durationMs: 90),
                AlarmTone(frequencyHz: 1300, durationMs: 90),
                AlarmTone(frequencyHz: 0, durationMs: 90),
                AlarmTone(frequencyHz: 1300, durationMs: 90),
                AlarmTone(frequencyHz: 0, durationMs: 90),
                AlarmTone(frequencyHz: 1300, durationMs: 90),
            ],
            tailSilenceMs: 400
        ),
        AlarmMelody(
            id: 2,
            nameKey: "melody_triplet",
            tones: [
                AlarmTone(frequencyHz: 800, durationMs: 130),
                AlarmTone(frequencyHz: 0, durationMs: 60),
                AlarmTone(frequencyHz: 900, durationMs: 130),
                AlarmTone(frequencyHz: 0, durationMs: 60),
                AlarmTone(frequencyHz: 1050, durationMs: 130),
            ],
            tailSilenceMs: 450
        ),
        AlarmMelody(
            id: 3,
            nameKey: "melody_rising",
            tones: [
                AlarmTone(frequencyHz: 600, durationMs: 90),
                AlarmTone(frequencyHz: 750, durationMs: 90),
                AlarmTone(frequencyHz: 900, durationMs: 90),
                AlarmTone(frequencyHz: 1050, durationMs: 90),
                AlarmTone(frequencyHz: 1200, durationMs: 110),
            ],
            tailSilenceMs: 400
        ),
        AlarmMelody(
            id: 4,
            nameKey: "melody_siren",
            tones: [
                AlarmTone(frequencyHz: 1100, durationMs: 300),
                AlarmTone(frequencyHz: 750, durationMs: 300),
                AlarmTone(frequencyHz: 1100, durationMs: 300),
                AlarmTone(frequencyHz: 750, durationMs: 300),
            ],
            tailSilenceMs: 250
        ),
    ]

    static func byId(_ id: Int) -> AlarmMelody {
        all.first { $0.id == id } ?? all[0]
    }

    static func displayName(_ key: String) -> String {
        switch key {
        case "melody_classic": return "Klasik Çalar Saat"
        case "melody_digital": return "Dijital Atış"
        case "melody_triplet": return "Acil Üçleme"
        case "melody_rising": return "Yükselen Alarm"
        case "melody_siren": return "Siren"
        default: return key
        }
    }
}
